ui/pages/confirmations/utils/token.ts

```diff
@@ -1,30 +1,89 @@
 import { memoize } from 'lodash';
 import { Hex } from '@metamask/utils';
+import { AssetsContractController } from '@metamask/assets-controllers';
 import { getTokenStandardAndDetails } from '../../../store/actions';
 
+export type TokenDetailsERC20 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC20Standard']>['getDetails']
+  >
+> & { decimalsNumber: number };
+
+export type TokenDetailsERC721 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC721Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetailsERC1155 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC1155Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetails =
+  | TokenDetailsERC20
+  | TokenDetailsERC721
+  | TokenDetailsERC1155;
+
 export const ERC20_DEFAULT_DECIMALS = 18;
 
+export const parseTokenDetailDecimals = (
+  decStr?: string,
+): number | undefined => {
+  if (!decStr) {
+    return undefined;
+  }
+
+  for (const radix of [10, 16]) {
+    const parsedDec = parseInt(decStr, radix);
+    if (isFinite(parsedDec)) {
+      return parsedDec;
+    }
+  }
+  return undefined;
+};
```

Comment:
Separated parseTokenDetailDecimals logic from fetchErc20Decimals to return undefined to handle unfound token detail decimals